### DISCUSSION TITLE
Is there a way to prevent many @overload?

### DISCUSSION THREAD

**Discussion by pbelskiy** [microsoft/pyright#6424] [Q&A] [answered]
Hello! I have a Python package with sync and async http adapters. I need correct type hints there.

It there any better way except using a lot of `@overload` for 100+ methods.
Or it's only the way, and I could only move out it signatures to some pyi files to increase readability?

```python
import asyncio

from typing import Any, Coroutine, overload, TypeVar, Generic, Union


ClientT = TypeVar('ClientT', bound=Union['JenkinsClient', 'AsyncJenkinsClient'])


class Builds(Generic[ClientT]):
    def __init__(self, jenkins: ClientT) -> None:
        self.jenkins = jenkins

    @overload
    def get(self: 'Builds[JenkinsClient]', name: str) -> str:
        ...

    @overload
    def get(self: 'Builds[AsyncJenkinsClient]', name: str) -> Coroutine[Any, Any, str]:
        ...

    def get(self, name: str) -> str | Coroutine[Any, Any, str]:
        return self.jenkins._request(name)


class AsyncJenkinsClient:
    def __init__(self) -> None:
        self.builds = Builds(self)

    async def _request(self, name: str) -> Any:
        return name


class JenkinsClient:
    def __init__(self) -> None:
        self.builds = Builds(self)

    def _request(self, name: str) -> Any:
        return name


async def main() -> None:
    async_client = AsyncJenkinsClient()
    result = await async_client.builds.get('async')
    print(result)

    client = JenkinsClient()
    result = client.builds.get('sync')
    print(result)


asyncio.run(main())
```

**Comment by erictraut** [comment, marked as answer]
For polymorphic functions like this, `@overload` is the go-to solution. 

Here's another approach that reduces the redundancy but isn't quite as clean.

Code sample in [pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwM4E8B2BjYECwAoAMyghDhhTGDQHM5RJY4BBNFAGjgGFoIBXGKgFMOAFQqCAagigcA4oLSCowDBwzIY%2BfACU4AXjhiwk6QAoARNvMBKLXgwAbZEjgAhXsAcATJKfmLlDABtbQBdawAufDgYuC9BAjgAfSSqYBgU0yRBBwIOACsFAGsqJAi4cwApYtLOB2AFeAAfZlRMarQStCQ6hrQYGzgAWgA%2BOAA5CEUovFi5uGzcgDpCztL9OFWupDs5%2BMTqQRgsnLy4NAQQQXKkGChrYbHtGfnYqCPeKDQ4dVvTbQ4iwIKxq3SWSXeAEdeII-hcrtZbHh8I5nK10BgOtteo0XrF9slUmh0pkgQ9RhMptdoq8FqclgAjDzeFwGdyeHxBbjEfhCIIsdjMViAu6hUInXJImkxZAYuIJZJQmF-IEceHXBZ3cljAV417vGCfb7quyopAuLG1eq46XyxIpNIZJISgjaynTW1zIGM5k%2BDbsllBW5QcVk3b4hUQwTQ2HHVXnS4a4Nu3Wet4fL4Jq52WWYO1wEAIKimN2TD2zWK5jBJRx9eAGJhtTGgnrW-ol23vJC8Bz1xAAdyL8CrNbbMB9HKQS0Ox3MVZstrAynbXZ7MClFZitcaG0t3Rx7aRc1XvY22-6E5Z06OFibC83cCXVGOJ-XOab2CWUF4aFMheLiL4EAA)

```python
import asyncio
from typing import Any, Coroutine, TypeVar, Generic, cast

R = TypeVar("R")

class Builds(Generic[R]):
    def __init__(self, jenkins: "JenkinsClient | AsyncJenkinsClient") -> None:
        self.jenkins = jenkins

    def get(self, name: str) -> R:
        return cast(R, self.jenkins._request(name))

class AsyncJenkinsClient:
    def __init__(self) -> None:
        self.builds = Builds[Coroutine[Any, Any, str]](self)

    async def _request(self, name: str) -> Any:
        return name

class JenkinsClient:
    def __init__(self) -> None:
        self.builds = Builds[str](self)

    def _request(self, name: str) -> Any:
        return name

async def main() -> None:
    async_client = AsyncJenkinsClient()
    result = await async_client.builds.get("async")
    print(result)

    client = JenkinsClient()
    result = client.builds.get("sync")
    print(result)

asyncio.run(main())
```

My only other suggestion is to rethink how you're abstracting this. It's a bit odd that you're passing in the client to the `Builds` and then saving a reference to `Builds` in the client. The circular dependency between these two classes is a design pattern that I would personally try to avoid. I think that's contributing to the challenge of typing this properly. If this is an existing library, and you need to maintain backward compatibility, then changing the abstraction isn't really an option.

**Reply by pbelskiy** [reply]
Thanks a lot!

**Comment by pbelskiy** [comment]
@erictraut

But actual return types are hidden from docs for example, and when code reading

<img width="733" alt="Screenshot 2023-11-13 at 23 05 36" src="https://github.com/microsoft/pyright/assets/8102549/289acd3e-b221-4fc1-8124-7ce059cbe652">
